File: crates/generator_android/src/android/dev_shell/canvas_layers.java
        private String doweActiveCanvasDrawMode() {
            Object value = drawModePath == null ? null : doweRead(drawModePath, null);
            return value == null ? drawMode : String.valueOf(value);
        }

        private ArrayList<Map<String, Object>> doweCanvasLayers() {
            ArrayList<Map<String, Object>> rows = new ArrayList<>();
            Object value = doweRead(layersPath, null);
            if (value instanceof List) for (Object entry : (List<?>) value) if (entry instanceof Map) rows.add(new HashMap<>((Map<String, Object>) entry));
            return rows;
        }

        private String doweCanvasLayerId(Map<String, Object> layer) { return layer.get("id") == null ? "" : String.valueOf(layer.get("id")); }

        private boolean doweCanvasLayerHit(Map<String, Object> layer, PointF point) {
            layer = doweBoundCanvasCommand(layer);
            float x = doweCanvasNumber(layer.get("x"), 0f), y = doweCanvasNumber(layer.get("y"), 0f);
            String type = String.valueOf(layer.get("type"));
            if ("circle".equals(type)) { float radius = doweCanvasNumber(layer.get("radius"), 0f) + Math.max(4f, doweCanvasNumber(layer.get("strokeWidth"), 1f)); return Math.hypot(point.x - x, point.y - y) <= radius; }
            float stroke = doweCanvasNumber(layer.get("strokeWidth"), 1f);
            if ("line".equals(type)) return doweCanvasSegmentDistance(point, new PointF(doweCanvasNumber(layer.get("x1"), 0f), doweCanvasNumber(layer.get("y1"), 0f)), new PointF(doweCanvasNumber(layer.get("x2"), 0f), doweCanvasNumber(layer.get("y2"), 0f))) <= Math.max(4f, stroke);
            if ("polyline".equals(type)) {
                List<PointF> points = doweCanvasLayerPoints(layer);
                if (points.size() == 1) return doweCanvasSegmentDistance(point, points.get(0), points.get(0)) <= Math.max(6f, stroke);
                for (int i = 1; i < points.size(); i++) if (doweCanvasSegmentDistance(point, points.get(i - 1), points.get(i)) <= Math.max(6f, stroke)) return true;
                return points.size() > 1 && Boolean.TRUE.equals(layer.get("closed")) && doweCanvasSegmentDistance(point, points.get(points.size() - 1), points.get(0)) <= Math.max(6f, stroke);
            }
            android.graphics.RectF bounds = doweCanvasLayerBounds(layer);
            return bounds != null && point.x >= bounds.left && point.x <= bounds.right && point.y >= bounds.top && point.y <= bounds.bottom;
        }

        private void doweCanvasLayerEvent(String action, String event, Map<String, Object> layer) {
            if (action == null) return;
            Map<String, Object> item = new HashMap<>(); item.put("event", event); item.put("layer", doweCopy(layer)); doweRunAction(action, item, this::doweRefreshCanvasPage);
        }

        private void doweUpdateCanvasLayer(PointF point, String kind) {
            if (layersPath == null) return;
            ArrayList<Map<String, Object>> rows = doweCanvasLayers();
            if ("down".equals(kind)) drawingMode = doweActiveCanvasDrawMode();
            String mode = drawingMode;
            if ("down".equals(kind)) {
                if ("select".equals(mode) || "erase".equals(mode)) {
                    Map<String, Object> selected = null;
                    for (int index = rows.size() - 1; index >= 0; index--) if (!doweCanvasLayerId(rows.get(index)).isEmpty() && doweCanvasLayerHit(rows.get(index), point)) { selected = rows.get(index); break; }
                    String id = selected == null ? "" : doweCanvasLayerId(selected);
                    if ("erase".equals(mode)) {
                        if (selected != null) { rows.remove(selected); doweWrite(layersPath, rows); if (id.equals(doweSelectedCanvasLayer())) { selectedLayerId = null; if (selectedPath != null) doweWrite(selectedPath, ""); } doweCanvasLayerEvent(onLayerRemove, "remove", selected); }
                    } else { selectedLayerId = id; if (selectedPath != null) doweWrite(selectedPath, id); doweCanvasLayerEvent(onLayerSelect, "select", selected); }
                    doweRefreshCanvasPage(); invalidate(); return;
                }
                boolean used = true;
                while (used) { used = false; String candidate = "layer-" + (++layerSequence); for (Map<String, Object> row : rows) if (candidate.equals(doweCanvasLayerId(row))) { used = true; break; } }
                String id = "layer-" + layerSequence; Map<String, Object> layer = new HashMap<>(); layer.put("id", id); layer.put("type", mode); layer.put("x", point.x); layer.put("y", point.y);
                if ("rect".equals(mode)) { layer.put("width", 0f); layer.put("height", 0f); layer.put("radius", 4f); }
                else if ("circle".equals(mode)) layer.put("radius", 0f);
                else { layer.put("type", "polyline"); ArrayList<Map<String, Object>> points = new ArrayList<>(); points.add(doweObject("x", point.x, "y", point.y)); layer.put("points", points); layer.put("closed", false); }
                layer.put("fill", "transparent"); layer.put("stroke", "primary"); layer.put("strokeWidth", 3f); rows.add(layer); doweWrite(layersPath, rows); drawingLayerId = id; drawingStart = point; invalidate(); return;
            }
            if (drawingLayerId == null || drawingStart == null) return;
            Map<String, Object> layer = null; int index = -1; for (int i = 0; i < rows.size(); i++) if (drawingLayerId.equals(doweCanvasLayerId(rows.get(i)))) { index = i; layer = rows.get(i); break; }
            if (layer == null) { drawingLayerId = null; drawingStart = null; return; }
            if ("move".equals(kind) || "up".equals(kind)) {
                if ("rect".equals(mode)) { layer.put("x", Math.min(drawingStart.x, point.x)); layer.put("y", Math.min(drawingStart.y, point.y)); layer.put("width", Math.abs(point.x - drawingStart.x)); layer.put("height", Math.abs(point.y - drawingStart.y)); }
                else if ("circle".equals(mode)) { layer.put("x", (drawingStart.x + point.x) / 2f); layer.put("y", (drawingStart.y + point.y) / 2f); layer.put("radius", Math.hypot(point.x - drawingStart.x, point.y - drawingStart.y) / 2f); }
                else {
                    List<Map<String, Object>> points = layer.get("points") instanceof List ? new ArrayList<>((List<Map<String, Object>>) layer.get("points")) : new ArrayList<>();
                    Map<String, Object> last = points.isEmpty() ? null : points.get(points.size() - 1);
                    if (last == null || doweCanvasNumber(last.get("x"), 0f) != point.x || doweCanvasNumber(last.get("y"), 0f) != point.y) points.add(doweObject("x", point.x, "y", point.y));
                    layer.put("points", points);
                }
                rows.set(index, layer); doweWrite(layersPath, rows);
            }
            if ("up".equals(kind)) {
                selectedLayerId = drawingLayerId; if (selectedPath != null) doweWrite(selectedPath, drawingLayerId);
                drawingLayerId = null; drawingStart = null;
                Map<String, Object> snapshot = (Map<String, Object>) doweCopy(layer);
                doweRunAction(onLayerAdd, doweObject("event", "add", "layer", doweCopy(snapshot)), () -> { doweRunAction(onLayerChange, doweObject("event", "change", "layer", doweCopy(snapshot)), this::doweRefreshCanvasPage); });
                doweRefreshCanvasPage();
            }
            else if ("cancel".equals(kind)) { rows.remove(index); doweWrite(layersPath, rows); drawingLayerId = null; drawingStart = null; }
            invalidate();
        }

        private void doweRemoveSelectedLayer() {
            if (layersPath == null) return;
            String selected = doweSelectedCanvasLayer(); if (selected == null || selected.isEmpty()) return;
            ArrayList<Map<String, Object>> rows = doweCanvasLayers(); Map<String, Object> removed = null; for (int index = rows.size() - 1; index >= 0; index--) if (selected.equals(doweCanvasLayerId(rows.get(index)))) { removed = rows.remove(index); break; }
            if (removed == null) return; doweWrite(layersPath, rows); selectedLayerId = null; if (selectedPath != null) doweWrite(selectedPath, ""); doweCanvasLayerEvent(onLayerRemove, "remove", removed); doweRefreshCanvasPage(); invalidate();
        }

        private void doweRunCanvasAction(String id, Map<String, Object> item) {
            doweRunAction(id, item, () -> {
                invalidate();
                if (layersPath != null && drawingLayerId == null && ("up".equals(item.get("kind")) || "cancel".equals(item.get("kind")))) doweRefreshCanvasPage();
            });
        }

        private String doweSelectedCanvasLayer() {
            Object value = selectedPath == null ? selectedLayerId : doweRead(selectedPath, null);
            return value == null ? "" : String.valueOf(value);
        }

        private void doweCancelCanvasLayer() {
            if (drawingLayerId != null && layersPath != null) {
                ArrayList<Map<String, Object>> rows = doweCanvasLayers();
                rows.removeIf(row -> drawingLayerId.equals(doweCanvasLayerId(row)));
                doweWrite(layersPath, rows);
            }
            drawingLayerId = null; drawingStart = null; drawingPointerId = null;
        }

        private boolean pageRefreshPending;

        private void doweRefreshCanvasPage() {
            if (pageRefreshPending) return;
            pageRefreshPending = true;
            post(() -> {
                pageRefreshPending = false;
                if (!mountedPath.equals(currentPath) || drawingLayerId != null) return;
                renderCurrentRoute(false);
            });
        }

        private List<PointF> doweCanvasLayerPoints(Map<String, Object> layer) {
            ArrayList<PointF> points = new ArrayList<>();
            if (layer.get("points") instanceof List) for (Object value : (List<?>) layer.get("points")) if (value instanceof Map) {
                Map<?, ?> point = (Map<?, ?>) value;
                points.add(new PointF(doweCanvasNumber(point.get("x"), 0f), doweCanvasNumber(point.get("y"), 0f)));
            }
            return points;
        }

        private float doweCanvasSegmentDistance(PointF point, PointF start, PointF end) {
            float dx = end.x - start.x, dy = end.y - start.y, length = dx * dx + dy * dy;
            float t = length == 0f ? 0f : Math.max(0f, Math.min(1f, ((point.x - start.x) * dx + (point.y - start.y) * dy) / length));
            return (float) Math.hypot(point.x - start.x - t * dx, point.y - start.y - t * dy);
        }

        private android.graphics.RectF doweCanvasLayerBounds(Map<String, Object> layer) {
            float x = doweCanvasNumber(layer.get("x"), 0f), y = doweCanvasNumber(layer.get("y"), 0f);
            String type = String.valueOf(layer.get("type"));
            if ("circle".equals(type)) { float r = doweCanvasNumber(layer.get("radius"), 0f); return new android.graphics.RectF(x - r, y - r, x + r, y + r); }
            if ("rect".equals(type) || "image".equals(type)) return new android.graphics.RectF(x, y, x + doweCanvasNumber(layer.get("width"), 0f), y + doweCanvasNumber(layer.get("height"), 0f));
            if ("text".equals(type)) {
                float size = doweCanvasNumber(layer.get("size"), 16f), width = Math.max(1, layer.get("text") == null ? 0 : String.valueOf(layer.get("text")).length()) * size * 0.6f;
                float left = x - ("center".equals(layer.get("align")) ? width / 2f : "end".equals(layer.get("align")) ? width : 0f);
                return new android.graphics.RectF(left, y - size, left + width, y);
            }
            List<PointF> points = doweCanvasLayerPoints(layer);
            if ("line".equals(type)) { points.add(new PointF(doweCanvasNumber(layer.get("x1"), 0f), doweCanvasNumber(layer.get("y1"), 0f))); points.add(new PointF(doweCanvasNumber(layer.get("x2"), 0f), doweCanvasNumber(layer.get("y2"), 0f))); }
            if (points.isEmpty()) return null;
            float left = points.get(0).x, top = points.get(0).y, right = left, bottom = top;
            for (PointF point : points) { left = Math.min(left, point.x); top = Math.min(top, point.y); right = Math.max(right, point.x); bottom = Math.max(bottom, point.y); }
            return new android.graphics.RectF(left, top, right, bottom);
        }

File: crates/generator_web/src/web/routes_and_css_collection/router_runtime/canvas_layers.js
function canvasLayerPath(canvas) {
  return canvas.dataset.doweCanvasLayerBind || "";
}
function canvasSelectedPath(canvas) {
  return canvas.dataset.doweCanvasSelected || "";
}
function canvasLayers(canvas, state, scope) {
  const value = readPath(state, canvasLayerPath(canvas), scope);
  return Array.isArray(value) ? value.slice() : [];
}
function canvasLayerId(layer) {
  return layer && layer.id != null ? String(layer.id) : "";
}
function canvasNextLayerId(canvas, layers) {
  let sequence = canvas.__doweLayerSequence || 0;
  const used = new Set(layers.map(canvasLayerId));
  do { sequence++; } while (used.has(`layer-${sequence}`));
  canvas.__doweLayerSequence = sequence;
  return `layer-${sequence}`;
}
function canvasSegmentHit(a, b, x, y, tolerance) {
  const ax = canvasNumber(a.x), ay = canvasNumber(a.y);
  const dx = canvasNumber(b.x) - ax, dy = canvasNumber(b.y) - ay;
  const length = dx * dx + dy * dy;
  const t = length ? Math.max(0, Math.min(1, ((x - ax) * dx + (y - ay) * dy) / length)) : 0;
  return Math.hypot(x - ax - t * dx, y - ay - t * dy) <= tolerance;
}
function canvasLayerBounds(layer) {
  const x = canvasNumber(layer.x), y = canvasNumber(layer.y);
  if (layer.type === "circle") {
    const radius = Math.max(0, canvasNumber(layer.radius));
    return { x: x - radius, y: y - radius, width: radius * 2, height: radius * 2 };
  }
  if (layer.type === "rect" || layer.type === "image")
    return { x, y, width: Math.max(0, canvasNumber(layer.width)), height: Math.max(0, canvasNumber(layer.height)) };
  if (layer.type === "text") {
    const size = Math.max(1, canvasNumber(layer.size, 16));
    const width = Math.max(1, String(layer.text || "").length) * size * 0.6;
    return { x: x - (layer.align === "center" ? width / 2 : layer.align === "end" ? width : 0), y: y - size, width, height: size };
  }
  const points = layer.type === "line"
    ? [{ x: layer.x1, y: layer.y1 }, { x: layer.x2, y: layer.y2 }]
    : Array.isArray(layer.points) ? layer.points : [];
  if (!points.length) return null;
  let left = Infinity, top = Infinity, right = -Infinity, bottom = -Infinity;
  for (const point of points) {
    left = Math.min(left, canvasNumber(point.x)); right = Math.max(right, canvasNumber(point.x));
    top = Math.min(top, canvasNumber(point.y)); bottom = Math.max(bottom, canvasNumber(point.y));
  }
  return { x: left, y: top, width: right - left, height: bottom - top };
}
function canvasLayerHit(layer, x, y) {
  if (!layer || typeof layer !== "object") return false;
  const tolerance = Math.max(layer.type === "polyline" ? 6 : 4, canvasNumber(layer.strokeWidth, 1));
  if (layer.type === "circle")
    return Math.hypot(x - canvasNumber(layer.x), y - canvasNumber(layer.y)) <= canvasNumber(layer.radius) + tolerance;
  if (layer.type === "line")
    return canvasSegmentHit({ x: layer.x1, y: layer.y1 }, { x: layer.x2, y: layer.y2 }, x, y, tolerance);
  if (layer.type === "polyline") {
    const points = Array.isArray(layer.points) ? layer.points : [];
    if (points.length === 1) return canvasSegmentHit(points[0], points[0], x, y, tolerance);
    for (let index = 1; index < points.length; index++)
      if (canvasSegmentHit(points[index - 1], points[index], x, y, tolerance)) return true;
    return !!layer.closed && points.length > 1 && canvasSegmentHit(points.at(-1), points[0], x, y, tolerance);
  }
  const bounds = canvasLayerBounds(layer);
  return !!bounds && x >= bounds.x && x <= bounds.x + bounds.width && y >= bounds.y && y <= bounds.y + bounds.height;
}
function canvasTopLayerAt(canvas, state, scope, x, y) {
  const layers = canvasLayers(canvas, state, scope);
  for (let index = layers.length - 1; index >= 0; index--)
    if (canvasLayerId(layers[index]) && canvasLayerHit(boundCanvasCommand(layers[index], state, scope), x, y))
      return { layer: layers[index], index };
  return null;
}
function canvasWriteLayers(canvas, state, layers) {
  const path = canvasLayerPath(canvas);
  if (path) writePath(state, path, layers);
}
function canvasRefreshLayers(canvas) {
  const view = getActiveView();
  if (view) renderReactive(view);
  scheduleDrawRender(canvas);
}
function canvasLayerEvent(canvas, event, layer) {
  const action = canvas.dataset[`doweCanvasOnLayer${event[0].toUpperCase()}${event.slice(1)}`];
  const item = JSON.parse(JSON.stringify({ event, layer: layer || {} }));
  if (action) {
    const view = getActiveView();
    const previous = canvas.__doweLayerEvents || Promise.resolve();
    canvas.__doweLayerEvents = previous.catch(() => {}).then(() => {
      if (getActiveView() === view) return runAction(action, { ...scopeFor(canvas), item });
    });
  }
}
function canvasSelectedId(canvas, state, scope) {
  const path = canvasSelectedPath(canvas);
  return path ? String(readPath(state, path, scope) || "") : canvas.__doweSelectedLayerId || "";
}
function canvasSetSelected(canvas, state, id) {
  canvas.__doweSelectedLayerId = id;
  const path = canvasSelectedPath(canvas);
  if (path) writePath(state, path, id);
}
function canvasSelectLayer(canvas, state, scope, layer) {
  canvasSetSelected(canvas, state, canvasLayerId(layer));
  canvasRefreshLayers(canvas);
  canvasLayerEvent(canvas, "select", layer);
}
function canvasRemoveLayer(canvas, state, scope, id) {
  const layers = canvasLayers(canvas, state, scope);
  const index = layers.findIndex(layer => id && canvasLayerId(layer) === id);
  if (index < 0) return;
  const removed = layers.splice(index, 1)[0];
  canvasWriteLayers(canvas, state, layers);
  if (canvasSelectedId(canvas, state, scope) === id) canvasSetSelected(canvas, state, "");
  canvasRefreshLayers(canvas);
  canvasLayerEvent(canvas, "remove", removed);
}
function canvasRemoveSelected(canvas, state, scope) {
  canvasRemoveLayer(canvas, state, scope, canvasSelectedId(canvas, state, scope));
}
function canvasConfiguredMode(canvas, state, scope) {
  const raw = canvas.dataset.doweCanvasDrawMode || "pen";
  const mode = canvas.dataset.doweCanvasDrawModeBinding === "true" ? String(readPath(state, raw, scope) || "pen") : raw;
  return ["pen", "rect", "circle", "select", "erase"].includes(mode) ? mode : "pen";
}
function drawCanvasSelection(ctx, canvas, state, scope) {
  const id = canvasSelectedId(canvas, state, scope);
  const layer = canvasLayers(canvas, state, scope).find(value => id && canvasLayerId(value) === id);
  const bounds = layer && canvasLayerBounds(boundCanvasCommand(layer, state, scope));
  if (!bounds) return;
  ctx.save();
  ctx.strokeStyle = canvasPaint("primary");
  ctx.lineWidth = 2;
  ctx.setLineDash([6, 4]);
  ctx.strokeRect(bounds.x - 5, bounds.y - 5, bounds.width + 10, bounds.height + 10);
  ctx.restore();
}
function canvasDrawGesture(canvas, state, scope, item, kind) {
  if (kind === "down") {
    if (canvas.__doweDrawGesture || !item.inside) return;
    const mode = canvasConfiguredMode(canvas, state, scope);
    const gesture = { id: item.id, mode, state, start: { x: item.x, y: item.y }, layer: null };
    canvas.__doweDrawGesture = gesture;
    if (mode === "select" || mode === "erase") {
      const hit = canvasTopLayerAt(canvas, state, scope, item.x, item.y);
      if (mode === "select") canvasSelectLayer(canvas, state, scope, hit?.layer);
      else if (hit) canvasRemoveLayer(canvas, state, scope, canvasLayerId(hit.layer));
      return;
    }
    const base = { id: canvasNextLayerId(canvas, canvasLayers(canvas, state, scope)), stroke: "primary", strokeWidth: 3 };
    gesture.layer = mode === "pen"
      ? { ...base, type: "polyline", points: [{ x: item.x, y: item.y }], closed: false }
      : mode === "rect"
        ? { ...base, type: "rect", x: item.x, y: item.y, width: 0, height: 0, fill: "transparent", radius: 4 }
        : { ...base, type: "circle", x: item.x, y: item.y, radius: 0, fill: "transparent" };
    if (canvasLayerPath(canvas)) canvasWriteLayers(canvas, state, [...canvasLayers(canvas, state, scope), gesture.layer]);
    else canvas.__doweDrawStrokes = [...(canvas.__doweDrawStrokes || []), gesture.layer];
  }
  const gesture = canvas.__doweDrawGesture;
  if (!gesture || gesture.id !== item.id) return;
  const { mode, start } = gesture;
  if (gesture.layer && (kind === "move" || kind === "up")) {
    let layer = { ...gesture.layer };
    if (mode === "pen") {
      const previous = layer.points.at(-1);
      if (!previous || previous.x !== item.x || previous.y !== item.y)
        layer.points = [...layer.points, { x: item.x, y: item.y }];
    }
    else if (mode === "rect") Object.assign(layer, { x: Math.min(start.x, item.x), y: Math.min(start.y, item.y), width: Math.abs(item.x - start.x), height: Math.abs(item.y - start.y) });
    else Object.assign(layer, { x: (start.x + item.x) / 2, y: (start.y + item.y) / 2, radius: Math.hypot(item.x - start.x, item.y - start.y) / 2 });
    gesture.layer = layer;
    if (canvasLayerPath(canvas))
      canvasWriteLayers(canvas, state, canvasLayers(canvas, state, scope).map(value => canvasLayerId(value) === layer.id ? layer : value));
    else canvas.__doweDrawStrokes = (canvas.__doweDrawStrokes || []).map(value => value.id === layer.id ? layer : value);
  }
  if (kind === "up" || kind === "cancel") {
    delete canvas.__doweDrawGesture;
    const layer = gesture.layer;
    if (layer && canvasLayerPath(canvas)) {
      if (kind === "cancel") canvasWriteLayers(canvas, state, canvasLayers(canvas, state, scope).filter(value => canvasLayerId(value) !== layer.id));
      else if (canvasLayers(canvas, state, scope).some(value => canvasLayerId(value) === layer.id)) {
        canvasSetSelected(canvas, state, layer.id);
        canvasLayerEvent(canvas, "add", layer);
        canvasLayerEvent(canvas, "change", layer);
      }
      canvasRefreshLayers(canvas);
    } else if (layer && kind === "cancel") canvas.__doweDrawStrokes = (canvas.__doweDrawStrokes || []).filter(value => value.id !== layer.id);
  }
  scheduleDrawRender(canvas);
}
